### ISSUE TITLE
Order of booleans in an && affects T.untyped vs T.nilable(Boolean)

### ISSUE THREAD

**Issue by professor** [sorbet/sorbet#6666]
While doing a "benign" code refactor, I noticed a new sorbet error for my code. What used to be a T.untyped now was a T.nilable(Boolean). In investigating what was happening, @alexevanczuk  and I noticed that the order of booleans was affecting the evaluation of the express.

Our work-around was to add `sig { returns(T::Boolean) }` to untyped methods. However, this unexpected behavior will trip others up and the error messages don't help the user to get to the work-around.

#### Input

[→ View on sorbet.run](https://sorbet.run/#x%20%3D%20T.let%281%2C%20T.untyped%29%0Ay%20%3D%20T.let%281%2C%20T.untyped%29%0AT.reveal_type%28true%20%26%26%20x%20%26%26%20y%29%0AT.reveal_type%28true%20%26%26%20x%20%26%26%20true%29%0A)

```ruby
x = T.let(1, T.untyped)
y = T.let(1, T.untyped)
T.reveal_type(true && x && y)
T.reveal_type(true && x && true)

```

#### Observed output

```
editor.rb:3: Left side of && condition was always truthy https://srb.help/7006
     3 |T.reveal_type(true && x && y)
                      ^^^^
    editor.rb:3: This condition was always truthy (TrueClass)
     3 |T.reveal_type(true && x && y)
                      ^^^^
  Got TrueClass originating from:
    editor.rb:3:
     3 |T.reveal_type(true && x && y)
                      ^^^^
  Note:
    If this is intentional, either delete the redundant code or restructure
    it to use T.absurd so that Sorbet can check for exhaustiveness.

editor.rb:3: Revealed type: T.untyped https://srb.help/7014
     3 |T.reveal_type(true && x && y)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  Got T.untyped originating from:
    editor.rb:1:
     1 |x = T.let(1, T.untyped)
            ^^^^^^^^^^^^^^^^^^^
    editor.rb:2:
     2 |y = T.let(1, T.untyped)
            ^^^^^^^^^^^^^^^^^^^
    editor.rb:3:
     3 |T.reveal_type(true && x && y)
                      ^^^^^^^^^

editor.rb:4: Left side of && condition was always truthy https://srb.help/7006
     4 |T.reveal_type(true && x && true)
                      ^^^^
    editor.rb:4: This condition was always truthy (TrueClass)
     4 |T.reveal_type(true && x && true)
                      ^^^^
  Got TrueClass originating from:
    editor.rb:4:
     4 |T.reveal_type(true && x && true)
                      ^^^^
  Note:
    If this is intentional, either delete the redundant code or restructure
    it to use T.absurd so that Sorbet can check for exhaustiveness.

editor.rb:4: Revealed type: T.nilable(T::Boolean) https://srb.help/7014
     4 |T.reveal_type(true && x && true)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  Got T.nilable(T::Boolean) originating from:
    editor.rb:1:
     1 |x = T.let(1, T.untyped)
            ^^^^^^^^^^^^^^^^^^^
    editor.rb:3:
     3 |T.reveal_type(true && x && y)
                      ^^^^^^^^^
    editor.rb:4:
     4 |T.reveal_type(true && x && true)
                      ^^^^^^^^^
    editor.rb:4:
     4 |T.reveal_type(true && x && true)
                                   ^^^^
Errors: 4
```

#### Expected behavior

I would expect the type to be consistently either untyped or nilable boolean.

**Comment by froydnj** [comment]
Thanks for the report.  I think this is a duplicate/amalgamation of #373 and #6259 ?

**Comment by paracycle** [comment]
@froydnj I believe so, but I think we can close this issue since the stated expected behaviour is not correct.

@professor Your expectation that type should be consistently either untyped of nilable boolean is not correct, since the logic operators are not commutative. Normally, one expects `&&` and `||` to be commutative, but short-circuiting breaks that assumption, and leads to the current behaviour you are seeing, which is the expected behaviour. Refer to my detailed note in https://github.com/sorbet/sorbet/issues/373#issuecomment-1154519194

The "fix", as suggested in #373 is to better report where the untyped or nilable is coming from in the error message, but the behaviour is correct.
